control-base/src/main/java/fi/nls/oskari/control/data/LayerAndGroupOrderHandler.java

```diff
@@ -0,0 +1,172 @@
+	 */
+	private HashMap<Integer, Integer> newGroupOrderMap = new HashMap<>();
+	/**
+	 * Variable to keep track of the new order of the layers.
+	 */
+	private HashMap<Integer, Integer> newLayerOrderMap = new HashMap<>(); 
+
+    public void setOskariMapLayerGroupService(final OskariMapLayerGroupService service) {
+        oskariMapLayerGroupService = service;
+    }
+    
+    public void setOskariLayerService(final OskariLayerService service) {
+    	oskariLayerService = service;
+    }
+
+    @Override
+    public void init() {
+        // setup service if it hasn't been initialized
+        if(oskariMapLayerGroupService == null) {
+            setOskariMapLayerGroupService(new OskariMapLayerGroupServiceIbatisImpl());
+        }
+        // setup service if it hasn't been initialized
+        if(oskariLayerService == null) {
+        	setOskariLayerService(new OskariLayerServiceIbatisImpl());
+        }
+        if(newGroupOrderMap == null) {
+        	newGroupOrderMap = new HashMap<>();
+        }
+        if(newLayerOrderMap == null) {
+        	newLayerOrderMap = new HashMap<>();
+        }
+    }
+
+    /**
+     * Handles updating the order and group of the given node i.e. layer or group.
+     * @param params
+     * @throws ActionException
+     */
+    @Override
+    public void handlePost(ActionParameters params) throws ActionException {
+        checkForAdminPermission(params);
+        log.debug("Updating layer/group order");
+        JSONObject orderJSON = getOrderJSON(params.getRequest());
+
+
+		try {
+			int parentID = orderJSON.getInt(KEY_PARENT);
+			JSONArray orders = orderJSON.getJSONArray(KEY_ORDERS);
+
+			// change main groups orders
+			if(parentID == -1){
+				for(int i=0;i<orders.length();i++){
+					JSONObject order = orders.getJSONObject(i);
+					int groupId = order.getInt(KEY_ID);
+					MaplayerGroup currentGroup = oskariMapLayerGroupService.find(groupId);
+					currentGroup.setOrderNumber(i);
+					oskariMapLayerGroupService.updateOrder(currentGroup);
+				}
+			} else {
+				for(int i=0;i<orders.length();i++) {
+					JSONObject order = orders.getJSONObject(i);
+					String type = order.getString(KEY_TYPE);
+					int id = order.getInt(KEY_ID);
+					if(TYPE_LAYER.equals(type)) {
+                        if(order.has(KEY_OLD_PARENT)) {
+                            oskariLayerService.updateGroup(id, order.getInt(KEY_OLD_PARENT), parentID);
+                        }
+					    oskariLayerService.updateOrder(id, parentID, i);
+					} else {
+						MaplayerGroup currentGroup = oskariMapLayerGroupService.find(id);
+						currentGroup.setOrderNumber(i);
+						oskariMapLayerGroupService.updateOrder(currentGroup);
+						if(order.has(KEY_OLD_PARENT)){
+						    oskariMapLayerGroupService.updateGroupParent(id,parentID);
+                        }
+					}
+				}
+			}
+
+
+
+        } catch (JSONException e) {
+        	log.warn(e);
+            throw new ActionException("Failed to read request!");
```

Comment:
This could be an `ActionParamsException` instead since the JSONException is thrown only if the JSON read from the request doesn't follow the required schema. The current exception message is also a bit misleading.

Reply:
Also the try/catch block is pretty big and formatting is all over the place :)